### PR TITLE
Add AES-CBC implementation (mbedtls)

### PR DESCRIPTION
**Pull request by omeg**

### CODE REVIEW

File: Pal/lib/crypto/adapters/mbedtls_adapter.c [hunk 1]
```diff
@@ -69,6 +71,100 @@ int lib_SHA256Final(LIB_SHA256_CONTEXT *context, uint8_t *output)
...
+                                    last_block, output + aligned_size);
+        if (ret != 0)
+            goto out;
+    } else {
+        /* Input size is a multiple of block length, encrypt everything
+           in one go. */
+        ret = mbedtls_aes_crypt_cbc(&aes, MBEDTLS_AES_ENCRYPT, input_len, iv,
+                                    input, output);
+        if (ret != 0)
+            goto out;
```
Comment (mkow) [inline Pal/lib/crypto/adapters/mbedtls_adapter.c]: `ret` is not PAL_* constant here. Same for other similar places in this code.

Comment (omeg) [reply Pal/lib/crypto/adapters/mbedtls_adapter.c]: Existing code does this (no special PAL* errors for crypto)

Comment (mkow) [reply Pal/lib/crypto/adapters/mbedtls_adapter.c]: Looks like a bug then, `mbedtls_*` functions return `MBEDTLS_*` constants, which collide numerically with `PAL_*`. Better to fix this IMO, because you'll get garbage instead of meaningful error messages in case of an error. Just return any meaningful `PAL_*` constant here.

Maybe it would also be worth to submit an issue or PR about other such places in the existing code?

File: Pal/lib/crypto/adapters/mbedtls_adapter.c [hunk 2]
```diff
@@ -74,94 +74,91 @@ int lib_SHA256Final(LIB_SHA256_CONTEXT *context, uint8_t *output)
...
-    mbedtls_aes_setkey_enc( &aes, key, key_len*8 );
+    mbedtls_aes_setkey_enc(&aes, key, key_len*8);
 
     /* padding */
-    uint64_t aligned_size = (input_len / 16) * 16;
-    uint64_t trailing_size = input_len - aligned_size;
-    uint64_t pad_size = 16 - trailing_size;
-    uint8_t last_block[16];
+    uint64_t pad_size = lib_AESGetPaddingSize(input_len);
+    uint8_t last_block[16] = { pad_size };
```
Comment (mkow) [inline Pal/lib/crypto/adapters/mbedtls_adapter.c]: For PKCS#5 padding you should fill the whole array with `pad_size`, not only the first byte.

File: Pal/lib/crypto/adapters/mbedtls_adapter.c [hunk 3]
```diff
@@ -74,94 +74,91 @@ int lib_SHA256Final(LIB_SHA256_CONTEXT *context, uint8_t *output)
...
 
 /* CBC decrypt, iv is assumed to be 16 bytes.
- * output_len must be equal to input_len and a multiple of block size (16). */
+ * input_len must be a multiple of block size (16).
+ * output is allocated, size == input_len. */
 int lib_AESCBCDecrypt(const uint8_t *key, uint64_t key_len,
                       uint8_t *iv,
                       const uint8_t *input, uint64_t input_len,
-                      uint8_t *output, uint64_t output_len)
+                      uint8_t **output)
```
Comment (mkow) [inline Pal/lib/crypto/adapters/mbedtls_adapter.c]: It would be handy to also return the output length (i.e. input_len - padding_size), so the external code doesn't have to calculate it.

Comment (omeg) [reply Pal/lib/crypto/adapters/mbedtls_adapter.c]: The decrypt function doesn't know the original input size though.

Comment (mkow) [reply Pal/lib/crypto/adapters/mbedtls_adapter.c]: It knows, that's the purpose of filling the padding with `pad_size` ;) Decrypted size is `input_size - decrypted_last_byte`.

File: Pal/lib/crypto/adapters/mbedtls_adapter.c [hunk 4]
```diff
@@ -23,10 +23,92 @@
...
+        case MBEDTLS_ERR_RSA_BAD_INPUT_DATA:
+            return -PAL_ERROR_CRYPTO_BAD_INPUT_DATA;
+
+        case MBEDTLS_ERR_RSA_OUTPUT_TOO_LARGE:
+            return -PAL_ERROR_CRYPTO_INVALID_OUTPUT_LENGTH;
+
+        case MBEDTLS_ERR_CIPHER_ALLOC_FAILED:
+        case MBEDTLS_ERR_DHM_ALLOC_FAILED:
+        case MBEDTLS_ERR_MD_ALLOC_FAILED:
+            return -PAL_ERROR_CRYPTO_ALLOC_FAILED;
```
Comment (mkow) [inline Pal/lib/crypto/adapters/mbedtls_adapter.c]: IMO PAL_ERROR_NOMEM would be better here, no need to introduce a new error for this.

File: Pal/src/pal_error.h [hunk 5]
```diff
@@ -52,46 +54,93 @@
...
+    { 1007, "[Crypto] Data misaligned" },
+    { 1008, "[Crypto] Invalid ASN.1 data" },
+    { 1009, "[Crypto] Memory allocation failed" },
+    { 1010, "[Crypto] Authentication failed" },
+    { 1011, "[Crypto] I/O error" },
+    { 1012, "[Crypto] Key generation failed" },
+    { 1013, "[Crypto] Invalid key" },
+    { 1014, "[Crypto] Public key operation failed" },
+    { 1015, "[Crypto] Private key operation failed" },
+    { 1016, "[Crypto] PKCS#1 verification failed" },
```
Comment (mkow) [inline Pal/src/pal_error.h]: How is this string related to `PAL_ERROR_CRYPTO_VERIFY_FAILED`?

File: Pal/src/pal_error.h [hunk 6]
```diff
@@ -52,46 +54,93 @@
...
+    { 1017, "[Crypto] RNG failed to generate data" },
+    { 1018, "[Crypto] Bignum operation failed" },
+};
+
+#define PAL_ERROR_NATIVE_COUNT   25
```
Comment (mkow) [inline Pal/src/pal_error.h]: Maybe it would be better to move it right after `#define PAL_ERROR_ADDRNOTEXIST         25`, so one won't break the code that easily when adding a new PAL_ERROR.

File: Pal/lib/crypto/adapters/mbedtls_adapter.c [hunk 7]
```diff
@@ -242,6 +243,7 @@ int lib_AESCBCDecrypt(const uint8_t *key, uint64_t key_len,
     if (ret != 0)
         goto out;
 
+    *output_len = input_len - (*output)[input_len-1];
```
Comment (mkow) [inline Pal/lib/crypto/adapters/mbedtls_adapter.c]: This will cause OOB read when `input_len` is 0. I'd also use `min(16u, (*output)[input_len-1])`, otherwise decrypting untrusted bytes can underflow `output_len`.

**Comment by mkow** [comment]
Looks good, you can also add a comment what's the origin of constants naming differences (e.g. `MBEDTLS_ERR_RSA_INVALID_PADDING` vs `PAL_ERROR_CRYPTO_INVALID_PADDING`).
